include/zephyr/bluetooth/hci_vs.h

```diff
@@ -216,9 +216,25 @@ struct bt_hci_evt_vs {
 } __packed;
 
 #define BT_HCI_EVT_VS_FATAL_ERROR              0x02
+
+#define BT_HCI_EVT_VS_ERROR_TYPE_SYSTEM 0x1
+#define BT_HCI_EVT_VS_ERROR_TYPE_CTRL_ASSERT 0x2
+struct bt_hci_vs_fatal_error_stack_frame {
+	uint32_t reason;
+#if defined(CONFIG_CPU_CORTEX_M)
+	uint32_t a1;
+	uint32_t a2;
+	uint32_t a3;
+	uint32_t a4;
+	uint32_t ip;
+	uint32_t lr;
+	uint32_t xpsr;
+#endif /* CONFIG_CPU_CORTEX_M */
```

Comment:
Not really sure how I feel about having a HCI event (even VS) to be have CPU type dependencies like this, but I guess it makes sense. 

Is this struct placed as the `data` in `bt_hci_evt_vs_fatal_error`? 

Would it make sense to already now introduce it as a `union`, so that it is easier to extend to other CPU types?

Reply:
> Is this struct placed as the data in bt_hci_evt_vs_fatal_error?

Yes it is used for `system fatal error` case as a `data` added to event.

>Would it make sense to already now introduce it as a union, so that it is easier to extend to other CPU types?

I think that is a good idea. For `raw` bytes/data like for a `string` user can go with `data` member, for other cases named member shall be provided.

Reply:
Since we already have the `type` defined, I guess we could use that to determine which struct to cast the data to, right? 

Does the data need to CPU specific, or do we just need to supply different formats, based on the type field? That would be similar to how HCI is defined in the first place as LTV structures.

Reply:
There are two options:
- we consider Host as being aware of an architecture of CPU where Controller runs
- or I'll add another common field to `bt_hci_evt_vs_fatal_error` with information about architecture. The CPU specific part may be also a union:
```
union {
        uint8_t cpu_data[0];
        struct bt_hci_evt_err_stack_frame sf_cortex_m;
};
```

I think that `cpu_data_length` is not required. The total length is already available in the event. The length of the data filed may be inferred from `type` and if available `cpu_type`.

Reply:
> Would it make sense to already now introduce it as a union, so that it is easier to extend to other CPU types?
and
> or I'll add another common field to `bt_hci_evt_vs_fatal_error` with information about architecture. The CPU specific part may be also a union

I do not want to create a unions there. If there is added support for another architecture that uses more data than Cortex-M stack frame, then the size of for Cortex-M will increase also. Additional bytes will be not used and they have to be zeroed anyway. I think it is better to construct the `data` and `cpu_data` by copying and have smallest possible event.

Reply:
> > 
> 
> I do not want to create a unions. If there is added support for another architecture that uses more data than Cortex-M stack frame, then the size of for Cortex-M will increase also. Additional bytes will be not used and they have to be zeroed anyway. I think it is better to construct the `data` and `cpu_data` by copying and have smallest possible event.

If we guard the union proper with e.g. `#if defined(CONFIG_CPU_CORTEX_M)`, then the size shouldn't be an issue. 

The point of defining the union was to add a reference to what structs could be put in `data`, e.g.

```c
struct bt_hci_vs_fata_error_cpu_data_cortex_m {
	uint32_t a1;
	uint32_t a2;
	uint32_t a3;
	uint32_t a4;
	uint32_t ip;
	uint32_t lr;
	uint32_t xpsr;
} __packed;

struct bt_hci_vs_fata_error_cpu_data_cortex_x {
	uint32_t a1;
	uint32_t a2;
	uint32_t a3;
	uint32_t a4;
} __packed;

union bt_hci_evt_vs_fatal_error_data {
#if defined(CONFIG_CPU_CORTEX_M)
	struct bt_hci_vs_fata_error_cpu_data_cortex_m cortex_m.
#endif /* CONFIG_CPU_CORTEX_M */
#if defined(CONFIG_CPU_CORTEX_X)
	struct bt_hci_vs_fata_error_cpu_data_cortex_x cortex_x.
#endif /* CONFIG_CPU_CORTEX_X */
};

struct bt_hci_evt_vs_fatal_error {
	uint8_t type;
	union bt_hci_evt_vs_fatal_error_data data;
} __packed;
```

where `CORTEX_X` is some other CPU.

Reply:
@Thalley that is fine from Controller perspective because it runs on a particular architecture. 
It is also fine from a conformance tests perspective because then we just read the received HCI events and interpret them.
However looking from Host perspective, I'm not sure if this is how it should be defined. Host may be not aware of architecture of a CPU the Controller is running on. Then the `cpu_data` should be interpreted dynamically. 
If we go that path, then in my humble opinion `bt_hci_evt_vs_fatal_error_data data;` should not be a union.

Reply:
That is a good point, however since this is an event, the host will never need to allocate it right? So the size of the `struct bt_hci_evt_vs_fatal_error` should not really matter for the host. 

I am fine with just keeping the `data` as a non-union though.

Reply:
Actually, since the host CPU may be different from the controller CPU, the design with the guards based on CPU type won't work. Please ignore my suggestion :D